src/mlpack/methods/lsh/lsh_search_impl.hpp

```diff
@@ -331,12 +333,424 @@ void LSHSearch<SortPolicy>::BaseCase(const size_t queryIndex,
+inline void perturbationExpand(std::vector<size_t>& A)
+{
+  size_t max = A[0];
+  for (size_t i = 1; i < A.size(); ++i)
+    if (A[i] > max)
+      max = A[i];
+  A.push_back(max + 1);
+}
+
+// Return true if perturbation set A is valid. A perturbation set is invalid if
+// it contains two (or more) actions for the same dimension or dimensions that
+// are larger than the queryCode's dimensions.
+inline bool perturbationValid(const std::vector<size_t>& A,
+                              const size_t numProj)
+{
+  // Stack allocation and initialization to 0 (bool check[numProj] = {0}) made
+  // some compilers complain, and std::vector might even be compressed (depends
+  // on implementation) so this saves some space.
+  std::vector<bool> check(numProj);
+
+  for (size_t i = 0; i < A.size(); ++i)
+  {
+    // Check that we only use valid dimensions. If not, vector is not valid.
+    if (A[i] >= 2 * numProj)
+      return false;
+
+    // Check that we only see each dimension once. If not, vector is not valid.
+    if (check[A[i] % numProj ] == 0)
+      check[A[i] % numProj ] = 1;
+    else
+      return false;
+  }
+  return true;
+}
+*/
+
+//Returns the score of a perturbation vector generated by perturbation set A.
+//The score of a pertubation set (vector) is the sum of scores of the
+//participating actions.
+inline double perturbationScore(const std::vector<bool>& A,
+                                const arma::vec& scores)
+{
+  double score = 0.0;
+  for (size_t i = 0; i < A.size(); ++i)
+    score += A[i] ? scores(i) : 0; //add scores of non-zero indices
+  return score;
+}
+
+// Inline function used by GetAdditionalProbingBins. The vector shift operation
+// replaces the largest element of a vector A with (largest element) + 1.
+inline bool perturbationShift(std::vector<bool>& A)
+{
+  size_t maxPos = 0;
+  for (size_t i = 0; i < A.size(); ++i)
+    if (A[i] == 1) // marked true
+      maxPos=i;
+  
+  if ( maxPos + 1 < A.size()) // otherwise, this is an invalid vector 
+  {
+    A[maxPos] = 0;
+    A[maxPos+1] = 1;
+    return true; // valid
+  }
+  return false; // invalid
+}
+
+// Inline function used by GetAdditionalProbingBins. The vector expansion
+// operation adds the element [1 + (largest_element)] to a vector A, where
+// largest_element is the largest element of A.
+inline bool perturbationExpand(std::vector<bool>& A)
+{
+  // Find the last '1' in A
+  size_t maxPos = 0;
+  for (size_t i = 0; i < A.size(); ++i)
+    if (A[i]) //marked true
+      maxPos = i;
+
+  if ( maxPos + 1 < A.size()) // otherwise, this is an invalid vector
+  {
+    A[maxPos+1] = 1;
+    return true;
+  }
+  return false;
+
+}
+
+// Return true if perturbation set A is valid. A perturbation set is invalid if
+// it contains two (or more) actions for the same dimension or dimensions that
+// are larger than the queryCode's dimensions.
+inline bool perturbationValid(const std::vector<bool>& A,
+                              const size_t numProj)
```

Comment:
Another small comment---we should name this function `PerturbationValid` (and the others should have a capital first letter too).  It might make more sense to put them inside of the `LSH` class just for the sake of organization, but whatever you want to do there is fine.